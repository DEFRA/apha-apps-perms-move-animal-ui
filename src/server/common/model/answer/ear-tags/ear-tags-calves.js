import { TextAreaAnswer } from '../text-area/text-area.js'

/** @import {TextAreaConfig} from '../text-area/text-area.js' */

/**
 * export @typedef {string} EarTagsData
 * @typedef {{ earTags: string }} EarTagsPayload
 */

/**
 * @augments {TextAreaAnswer<EarTagsPayload>}
 */
export class EarTagsCalvesAnswer extends TextAreaAnswer {
  /** @type {TextAreaConfig} */
  static config = {
    payloadKey: 'earTags',
    isPageHeading: false,
    hint: 'You need to enter each ear tag number on a separate line',
    rows: 15,
    validation: {
      maxLength: {
        value: 5000,
        message: 'Your answer must be no longer than 5000 characters'
      },
      empty: {
        message:
          'Enter the ear tag numbers of the calves under 42 days old you are planning to move'
      }
    }
  }
}
