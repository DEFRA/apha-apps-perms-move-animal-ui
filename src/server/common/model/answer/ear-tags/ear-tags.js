import { TextAreaAnswer } from '../text-area/text-area.js'

/** @import {TextAreaConfig} from '../text-area/text-area.js' */

/**
 * export @typedef {string} EarTagsData
 * @typedef {{ earTags: string }} EarTagsPayload
 */

/**
 * @augments {TextAreaAnswer<EarTagsPayload>}
 */
export class EarTagsAnswer extends TextAreaAnswer {
  /** @type {TextAreaConfig} */
  static config = {
    payloadKey: 'earTags',
    rows: 10,
    validation: {
      maxLength: {
        value: 5000,
        message: 'Your answer must be no longer than 5000 characters'
      },
      empty: {
        message:
          'Enter the ear tag numbers of the animals you are planning to move'
      }
    }
  }
}
