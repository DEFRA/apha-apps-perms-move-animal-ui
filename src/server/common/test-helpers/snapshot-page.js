import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
