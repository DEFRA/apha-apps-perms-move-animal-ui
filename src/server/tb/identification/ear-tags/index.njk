{% extends 'layouts/questions.njk' %}

{% block heading %}
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <h1 class="govuk-heading-l">{{ heading }}</h1>
    </div>
  </div>
{% endblock %}

{% block beforequestion %}
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <p>
        You need to identify the animals using the official ear tags bought from
        an approved supplier.
      </p>
      <p>
        The ear tag must include the country code, herd number and six-digit
        official ear tag number. For example, UK12345678910.
      </p>
    </div>
  </div>
{% endblock %}
