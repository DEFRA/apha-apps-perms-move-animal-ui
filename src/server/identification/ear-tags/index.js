import { QuestionPageController } from '../../common/controller/question-page-controller/question-page-controller.js'
import { EarTagsAnswer } from '../../common/model/answer/ear-tags/ear-tags.js'
import { QuestionPage } from '../../common/model/page/question-page-model.js'
import { identificationSummaryPage } from '../check-answers/index.js'

export class EarTagsPage extends QuestionPage {
  urlPath = '/identification/enter-ear-tags'
  sectionKey = 'identification'
  question = 'Enter the ear tag numbers of the animals you are planning to move'
  questionKey = 'earTags'
  Answer = EarTagsAnswer

  /** @param {EarTagsAnswer} _answer */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  nextPage(_answer) {
    return identificationSummaryPage
  }
}

export const earTagsPage = new EarTagsPage()

/**
 * @satisfies {ServerRegisterPluginObject<void>}
 */
export const earTags = new QuestionPageController(earTagsPage).plugin()

/**
 * @import { ServerRegisterPluginObject } from '@hapi/hapi'
 */
