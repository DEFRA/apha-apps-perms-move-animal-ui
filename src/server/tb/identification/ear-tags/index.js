import { TbQuestionPageController } from '../../question-page-controller.js'
import { EarTagsAnswer } from '../../../common/model/answer/ear-tags/ear-tags.js'
import { QuestionPage } from '../../../common/model/page/question-page-model.js'
import { identificationSummaryPage } from '../summary/index.js'

const customHeading =
  'Official ear tag numbers for animals 42 days old or older'

export class EarTagsPage extends QuestionPage {
  urlPath = '/identification/enter-ear-tags'
  sectionKey = 'identification'
  question = 'Enter the ear tag numbers for these animals'
  questionKey = 'earTags'
  Answer = EarTagsAnswer

  get heading() {
    return customHeading
  }

  view = 'tb/identification/ear-tags/index'

  nextPage() {
    return identificationSummaryPage
  }
}

export const earTagsPage = new EarTagsPage()

/**
 * @satisfies {ServerRegisterPluginObject<void>}
 */
export const earTags = new TbQuestionPageController(earTagsPage).plugin()

/**
 * @import { ServerRegisterPluginObject } from '@hapi/hapi'
 */
