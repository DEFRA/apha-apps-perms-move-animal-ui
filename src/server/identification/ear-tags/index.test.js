import { EarTagsAnswer } from '../../common/model/answer/ear-tags/ear-tags.js'
import { describePageSnapshot } from '../../common/test-helpers/snapshot-page.js'
import { IdentificationSummaryPage } from '../check-answers/index.js'
import { earTagsPage, EarTagsPage } from './index.js'

const sectionKey = 'identification'
const question =
  'Enter the ear tag numbers of the animals you are planning to move'
const questionKey = 'earTags'
const view = 'common/model/page/question-page.njk'
const pageUrl = '/identification/enter-ear-tags'

describe('EarTagsPage', () => {
  const page = new EarTagsPage()

  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct view', () => {
    expect(page.view).toBe(view)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(EarTagsAnswer)
  })

  it('nextPage should return check answers page', () => {
    const answer = new EarTagsAnswer({ earTags: 'some-text' })
    const nextPage = page.nextPage(answer)
    expect(nextPage).toBeInstanceOf(IdentificationSummaryPage)
  })

  it('should export page', () => {
    expect(earTagsPage).toBeInstanceOf(EarTagsPage)
  })

  describePageSnapshot({
    describes: 'EarTagsPage.content',
    it: 'should render expected response and content',
    pageUrl
  })
})
