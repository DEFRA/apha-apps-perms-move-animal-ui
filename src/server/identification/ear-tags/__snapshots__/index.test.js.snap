// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EarTagsPage EarTagsPage.content should render expected response and content 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <h1 class="govuk-heading-l">Official ear tag numbers for animals 42 days old or older</h1>
    </div>
  </div>
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <p>
        You need to identify the animals using the official ear tags bought from
        an approved supplier.
      </p>
      <p>
        The ear tag must include the country code, herd number and six-digit
        official ear tag number. For example, UK12345678910.
      </p>
    </div>
  </div>
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<fieldset class="govuk-fieldset">
  <div class="govuk-form-group">
  <label class="govuk-label govuk-label--m" for="earTags">
    Enter the ear tag numbers for these animals
  </label>
  <div id="earTags-hint" class="govuk-hint">
    You need to enter each ear tag number on a separate line
  </div>
  <textarea class="govuk-textarea" id="earTags" name="earTags" rows="15" aria-describedby="earTags-hint"></textarea>
</div>


</fieldset>
            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
